Add some security checks when handling a HTTP connection.

jgraham opened the pull request:
Check the Host and Origin headers for the incoming connection to
verify the connection is allowed.

The language is intended to allow the specific behaviour to be largely
implementation defined, whilst recommending a default
behaviour that prevents CSRF-type attacks (reject host headers that
aren't an IP address or the server hostname, reject any requests with
an origin header).

Hopefully adding this text will ensure that implementations consider
the security issues accepting a connection, even though it's not
possible to give precise requirements that apply to all
implementations.


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/w3c/webdriver/pull/1634.html" title="Last updated on Dec 3, 2021, 4:36 PM UTC (68f0fb1)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webdriver/1634/1917e89...68f0fb1.html" title="Last updated on Dec 3, 2021, 4:36 PM UTC (68f0fb1)">Diff</a>


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/w3c/webdriver/pull/1634.html" title="Last updated on Mar 18, 2022, 3:40 PM UTC (a26a826)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webdriver/1634/cfd3d52...a26a826.html" title="Last updated on Mar 18, 2022, 3:40 PM UTC (a26a826)">Diff</a>

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/w3c/webdriver/1634)
<!-- Reviewable:end -->

jgraham committed:
Add some security checks when handling a HTTP connection.

Check the Host and Origin headers for the incoming connection to
verify the connection is allowed.

The language is intended to allow the specific behaviour to be largely
implementation defined, whilst recommending a default
behaviour that prevents CSRF-type attacks (reject host headers that
aren't an IP address or the server hostname, reject any requests with
an origin header).

Hopefully adding this text will ensure that implementations consider
the security issues accepting a connection, even though it's not
possible to give precise requirements that apply to all
implementations.

jgraham commented:
The BiDi version of this is https://github.com/w3c/webdriver-bidi/pull/155

whimboo commented:
CC @sadym-chromium, @bwalderman

sadym-chromium commented on index.html:
```diff
@@ -428,8 +428,52 @@ <h3>Processing model</h3>
+  <li><p><p>If <var>request</var> has an <a>Origin header</a>,
+  let <var>origin</var> be the value of that header. Otherwise
+  let <var>origin</var> be null.</p></li>
+
```
When WebDriver server is run in a Docker container, the `Host header` is always different from the `host of the HTTP server`. Current spec includes this specific to "**_or to another host the implementation has been configured to allow._**", but maybe it would make sense to extract it into the specific step, so that the need of such a configuration will be more explicit?

```suggestion

 <li><p>Let <var>implementation allowed hosts rule</var> be the configured by the implementation rule determines if the given host allowed or not. Otherwise let <var>implementation allowed hosts rule</var> be the rule matching nothing.</p></li>
  
```

jgraham replied:
I think it's hard to be very specific here because the details of what's required depend a lot on the setup you want to support. For example it's possible to support the docker setup by making the host of the HTTP server the same as the host of the container. But it's also possible to support it by e.g. binding to an IP address, in which case the server just has to know it should accept requests with the specific host header that the conatiner will have. So I've tried to write some general guidance in the security section which suggests you probably want to provide some configuration options to support this kind of scenario, but without saying too much about exactly what they need to be.

sadym-chromium commented on index.html:
```diff
@@ -428,8 +428,52 @@ <h3>Processing model</h3>
+    <li><p>The result of <a>host parsing</a> the <code>uri-host</code>
+        part of <var>host</var> is not an <a>IP address</a>,
+        a <a>domain</a> identical to the
+        <a>host</a> of the HTTP server or to another <a>host</a> the
+        implementation has been configured to allow.</p></li>
```
```suggestion
        <a>host</a> of the HTTP server or 
        does not match the <var>implementation allowed hosts rule</var>.</p></li>
```

sadym-chromium commented on index.html:
```diff
@@ -428,8 +428,52 @@ <h3>Processing model</h3>
+  configured to allow, then stop running these steps and act as if the
+  requested service is not available.</p>
+
+  <p class=note>Rejecting connections with unexpected values in
+  the <a>Origin header</a> is necessary to prevent untrusted websites
+  from establishing a WebDriver session.</p></li>
```
Or maybe add note about Docker and some other edge cases here?

whimboo commented:
@jgraham I assume the summary of this issue should say `WebDriver` and not `WebSocket` right? 

Beside that please also have a look at issues #1578 and #1643. Both are related and I hope that we could fix them with your PR. Thanks.

jgraham committed:
Use undefined rather than null as the defaut header value.

This avoids confusion with null as a string value of the Origin header

jgraham committed:
Disallow CORS-safelisted Content-Type headers.

This prevents e.g. websites creating forms that can start a local WebDriver
session in a browser, even if there's no Origin header

jgraham committed:
Add some extra detail about allowing external network connections.

This doesn't add normative requirements, but encourages implementors to ship with secure
defaults, whilst mentioning that there are real-world use cases which require remote ends
to accept connections coming from somewhere other than localhost.

jgraham commented:
I've updated this a bit; I think it now covers all the points in the linked issues apart from:
* I didn't mention `AF_LOCAL` sockets. It definitely seems reasonable to support local-only communication as well as `AF_INET*` sockets, but I'm not sure in practice if they're usable in current clients ("local ends"). 
* I didn't mention HTTP Auth, although I did mention some kind of authorization being encouraged in cases where the "remote end" is really a publically accessible service. In practice I don't know what we could suggest which would add meaingful additional security and be compatible with common use cases like running tests in a CI system where the loal and remote ends are both on the same machine.

In any case I think if there's more to do for those points we should consider them in their own issues / PRs and limit this one to just the header validation requirements for accepting incomping requests.

whimboo commented:
@sadym-chromium in case you have some available time soon maybe you could have a look at this PR again and if your comments have been addressed or clarified? Thanks.

randomstuff commented on index.html:
```diff
@@ -428,8 +428,72 @@ <h3>Processing model</h3>
+        a <a>domain</a> identical to the
+        <a>host</a> of the HTTP server or to another <a>host</a> the
+        implementation has been configured to allow.</p></li>
+    <li><p>The <code>port</code> part of <var>host</var> is present
```
I don't believe verifying the port of the Host header adds any significant security (especially with respect to DNS rebinding attacks) (?). It is probably not necessary to require this verification.
